### PR TITLE
refactor: remove helper functions from `Transaction` and use `ShareableDatabase` instead.

### PR DESCRIPTION
**Pull request by joshieDo**
* `Transaction` takes a `ChainSpec` on its constructor
* `Transaction` can return a `&ShareableDatabase<&DB>`, which allows every stage to access the provider traits/functions (take a look at `crates/stages/src/stages/execution.rs`
* Removes some helper functions from `Transaction`, replacing its usage with the provider ones.

Looking for feedback on if it's the way to proceed.

One current issue is that this way, there's no transaction reuse ... leading to the opening and closing of many transactions. Same issue as here https://github.com/paradigmxyz/reth/issues/2871 .

### CODE REVIEW

File: crates/consensus/beacon/src/engine/mod.rs [hunk 1]
```diff
@@ -1270,7 +1270,14 @@ mod tests {
...
 
     fn insert_blocks<'a, DB: Database>(db: &DB, mut blocks: impl Iterator<Item = &'a SealedBlock>) {
-        let mut transaction = Transaction::new(db).unwrap();
+        let chain_spec = Arc::new(
+            ChainSpecBuilder::default()
+                .chain(MAINNET.chain)
+                .genesis(MAINNET.genesis.clone())
+                .paris_activated()
+                .build(),
+        );
```
Comment (gakonst) [inline crates/consensus/beacon/src/engine/mod.rs:1279]: Should this be reused across all calls?

File: crates/stages/src/stages/execution.rs [hunk 2]
```diff
@@ -88,9 +89,20 @@ impl<EF: ExecutorFactory> ExecutionStage<EF> {
...
+            .header_by_number(block_number)?
+            .ok_or_else(|| ProviderError::HeaderNotFound(block_number.into()))?;
+
+        let td = shareable_db
+            .header_td_by_number(block_number)?
+            .ok_or(ProviderError::TotalDifficultyNotFound { number: block_number })?;
+
+        let ommers = shareable_db
+            .ommers(block_number.into())?
+            .ok_or(ProviderError::BlockOmmers { number: block_number })?;
```
Comment (gakonst) [inline crates/stages/src/stages/execution.rs:103]: Can we not make the first call return a DbTx, and then call the methods over extension traits?

**Review by gakonst — COMMENTED** [review]
I like the direction of removing functions from the Transaction struct. IMO the right path here is to figure out how to add the ability to open and manage multiple read txs and a single write tx from it, and then implement extension traits for DbTx and DbTxMut. That will allow us to reuse the same transaction definitely inside a single RPC handler call, and we can maybe have a way to manage how long a DbTx is held open

**Comment by joshieDo** [comment]
closing as per the sync

**Pull request closed without merging**